src/test/java/lambda/part3/exercise/Mapping.java

```diff
@@ -405,83 +398,63 @@ public boolean forNext(Consumer<R> c) {
             ReachIterable<T> self = this;
 
             return new ReachIterable<R>() {
-                @Override
-                public void resetIteration() {
-                    self.resetIteration();
-                }
+
+                private ReachIterable<R> currentIterable =
+                        ReachIterable.from(Collections.emptyList());
+                private Consumer<T> getNewIterable =
+                        (T t) -> currentIterable = ReachIterable.from(f.apply(t));
 
                 @Override
                 public boolean forNext(Consumer<R> c) {
-                    return self.forNext(t ->
-                            f.apply(t).forEach(c)
-                    );
+                    boolean result = currentIterable.forNext(c);
+
+                    if (!result) {
+                        boolean isNotEnd = self.forNext(getNewIterable);

```

Comment:
Incorrect. Add test like this: `Arrays.asList("", "", "", "a")` -> `flatMap(getChars).force()`.